packages/firestore/src/api/credentials.ts

```diff
@@ -199,38 +197,35 @@ export class FirebaseCredentialsProvider implements CredentialsProvider {
       'getToken cannot be called after listener removed.'
     );
 
-    // Take note of the current value of the tokenCounter so that this method
-    // can fail (with an ABORTED error) if there is a token change while the
-    // request is outstanding.
-    const initialTokenCounter = this.tokenCounter;
     const forceRefresh = this.forceRefresh;
     this.forceRefresh = false;
 
     if (!this.auth) {
       return Promise.resolve(null);
     }
 
+    this.lastToken = null;
     return this.auth.getToken(forceRefresh).then(tokenData => {
-      // Cancel the request since the token changed while the request was
-      // outstanding so the response is potentially for a previous user (which
-      // user, we can't be sure).
-      if (this.tokenCounter !== initialTokenCounter) {
+      if (this.lastToken) {
+        // If we received a new token while this callback was pending (via
```

Comment:
Question: what is the reason to prefer `lastToken` to `tokenData` here?

Reply:
I would like you to verify this - but I think this matches the previous behavior. If we receive a token update via our callback listener, we ignore the token that is returned from the `getToken` call.

FWIW, while I think this is the cleaner solution to our token handling, I think the risk of it probably means we should not merge it. After all, we now have evidence that the problem we are trying to solve also exists in at least one other place (and maybe only that place).

Maybe @wilhuff wants a distraction from his current work and take a peak as well.

Reply:
The reason this is complicated is that we need to subscribe to auth to get notified when the identity changes, so that we can trigger user changes in the local store.

The reason we ignored the token in that callback was because when remote store wants to restart streams it it's going to trigger a call to getToken. This restart might happen significantly later, because there may not have been active streams when the identity changed.

The central question we need to answer is whether or not the token listener will call us before the current access token expires. If it does, it seems like we could just always use the token supplied by the token listener, no? If we did that this could be much simpler.